src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/PackageCollection.cs

```diff
@@ -54,5 +54,21 @@ public static async Task<PackageCollection> FromProjectsAsync(IEnumerable<IProje
 
             return new PackageCollection(packages);
         }
+
+        public static async Task<PackageCollection> FromProjectsTransitiveAsync(IEnumerable<IProjectContextInfo> projects, CancellationToken cancellationToken)
+        {
+            // Read transitive package references from all package reference style projects.
+            IEnumerable<Task<IReadOnlyCollection<IPackageReferenceContextInfo>>>? tasks = projects
+                .Select(project => project.GetTransitivePackagesAsync(cancellationToken).AsTask());
+            IEnumerable<IPackageReferenceContextInfo>[]? transitivePackageReferences = await Task.WhenAll(tasks);
```

Comment:
Methods returning `IEnumerable<T>` is one of many pet-peeves of mine. They are lazily evaluated, and the worst is when a method returns the result of a LINQ query, because if the `IEnumerable` is enumerated multiple times, the entire LINQ expression is re-evaluated multiple times. Many people don't realize this perf problem. This can actually be a correctness error in some scenarios, although very unlikely in NuGet.Client's code. Resharper has an analyzer rule that warns on multiple enumerations. I really wish Roslyn did too, although that's not what's going on here.

Unfortunately changing the return types of NuGet SDK projects under `src\NuGet.Core\*` would be breaking changes, so not realistic to change. But I strongly encourage us to comment on any PR making changes to any `src\NuGet.Clients\*` project that has methods or properties returning `IEnumerable<T>` to be changed. `IEnumerable<T>` as an input parameter is great when the input is only used once in a `foreach` loop. But methods/properties should never return that type, unless it's designed to be part of a LINQ expression (`public static IEnumerable<T> DoSomething<T>(this IEnumerable<T> input) { /* whatever */ }`)

Reply:
Thanks for the explanation. 
The code that I added follows the same pattern as FromProjectsAsync. Should I leave this as is, or make changes to both?

Reply:
Only modify your own code. If you happen to modify some existing code and notice the problem, then it's a good opportunity to fix it. But don't make your PR too much larger than it needs to be.

Reply:
Modified this to return both installed and transitive packages as (IReadOnlyCollection<IPackageReferenceContextInfo>, IReadOnlyCollection<IPackageReferenceContextInfo>)[].